### PR TITLE
Wmattei/request id with custom header

### LINKED ISSUE #8: Could the requestId middleware support custom header key?
Great lib folks.

I want to start using this on one of my projects, and I was wondering if the `requestId` built-in middleware could support different header key than `X-Request-ID`

If so, I would love to raise a PR doing it (:

### PR DESCRIPTION
**Pull request by wmattei**
This PR is related to #8.

I am creating a new middleware similar to `RequestID` but it gets a parameter where the developer can tell which header key will be used to get the requestId.

Also, it won't generate a new requestId if the header is no present as this assumes that there will always be a header.

Apart of that, I've also created a test file for this new middleware.

### CODE REVIEW

**Comment by wmattei** [comment]
Hey @saurabh0719 

As there are no contributing guides, I've just added a simple description to the PR. Please let me know what you think about it, and if there is something you would like differently please let me know.

BTW I used [this middleware from gin](https://github.com/gin-contrib/requestid) as inspiration

File: middleware/requestID.go [hunk 1]
```diff
@@ -68,6 +67,19 @@ func RequestID(next http.Handler) http.Handler {
 	})
 }
 
+// RequestIDWithCustomHeaderStrKey is a middleware that injects a request ID into the context of each
+// request. Different from RequestID, this middleware uses a custom header key to get the request ID,
+// and does not generate a new request ID if the custom header key is not present.
+func RequestIDWithCustomHeaderStrKey(headerStrKey string) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			requestID := req.Header.Get(headerStrKey)
```
Comment (saurabh0719) [inline middleware/requestID.go:76]: @wmattei 

In this particular line, If the client has _not_ set the Request ID header OR for an empty Request ID for that particular request, the `requestID` here will be an empty quote "". Having worked with only server generated request IDs, I'm a bit curious about the generally accepted standards here - 

- Should the server revert with a `400` as the client has forgotten to set the header? 
- What about if the Header is present but is empty?

**Comment by saurabh0719** [comment]
@wmattei 

I have a doubt in the following line - 
```go
requestID := req.Header.Get(headerStrKey)
```

The `Get` method will return "" if the RequestID Header is NOT present as well as if its present but empty. Having worked only with Server generated Request IDs, I'm curious what the accepted standard is for handling this case - 

- Should we return with a `400` if the Header itself is not present? 
- What if it's present but empty? As of now I see you've mentioned that it should not be generated on the server, and that seems agreeable. 

Just wondering that if the developer has explicitly requested a Request Key from the client by using this middleware, should it be enforced or not? And why?

**Comment by saurabh0719** [comment]
@wmattei 

Also I'd like to propose a tiny change to the naming conventions you've used - I'd prefer to remove the `Str` substring from the variables/functions/test as it's already evident from the function argument and the go doc.

Eg. Change `RequestIDWithCustomHeaderStrKey` to `RequestIDFromCustomHeader` and `headerStrKey` to `headerKey`

**Comment by wmattei** [comment]
> Just wondering if the developer has explicitly requested a Request Key from the client by using this middleware, should it be enforced or not? And why?

That is a good question. I would imagine that as I am explicitly using the middleware with custom naming, I would make sure my front end is always sending it. 

But by thinking better about it. What if I want to use a custom key (because that is what the front end agreed to send) but still generate it in case the key is not there?

Maybe we should keep generating it.
I am happy to do that

What do you think?

Also in the future, we might do something even more complex. e.g what if the developer wants the `requestId` to do a concatenation of multiple values, like `host+ip+protocol+body-value+header-value` etc...

**Comment by wmattei** [comment]
> Also I'd like to propose a tiny change to the naming conventions you've used - I'd prefer to remove the Str substring from the variables/functions/test as it's already evident from the function argument and the go doc.

Sure thing will do that tonight!

**Comment by saurabh0719** [comment]
@wmattei 

So one thing that we shouldn't do for sure is generate it on the server, as that defeats the entire purpose of client generated request ID keys and can get confusing (as the key generation logic is split between client and server).

However I feel there should be a check to see if the header is actually present AND not empty (a simple empty string check will do), as the server is explicitly requesting for it. If it's empty, I think it makes sense to return a `400` to avoid calling further middleware like the `Logger` which may depend on this. 

I feel in most production usecases, the client will always send a Request ID with the custom header. 

Any casual usecase can use their own custom middleware or just skip Request IDs all together.

**Comment by saurabh0719** [comment]
@wmattei on further thought, to avoid a mismatch in behaviour between the regular RequestID middleware and the Custom header one, I think it's okay to generate it on the server side if the client does not send one, as you earlier suggested. 

My thinking behind this is 
- Same behaviour of both functions makes it easier to understand for a user of this package 
- Request IDs are often sent back to the client in the response, so if the client hasn't explicitly set one, the server can generate it anyways and send it back to the client to store for any future queries/debug requests.

**Comment by saurabh0719** [comment]
Will merge it after you've made the edits. :)

**Comment by wmattei** [comment]
Sounds great!
Will work on that later today. And will also update the README.md file which I forgot!

**Comment by wmattei** [comment]
Hey @saurabh0719 all done!

I decided not to update the README.md because I realized that you only mention the essential middlewares there. Whereas I don't assume this to be an essential one (:

**Comment by saurabh0719** [comment]
Thanks @wmattei , merging it now. I'll update the README, there's work to be done on that anyways.

**Pull request merged**